R/regex_subset_linter.R

```diff
@@ -0,0 +1,88 @@
+#' Require usage of direct methods for subsetting strings via regex.
+#'
+#' Using `value = TRUE` in [grep()] returns the subset of the input that matches
+#'   the pattern, e.g. `grep("[a-m]", letters, value = TRUE)` will return the
+#'   first 13 elements (`a` through `m`).
+#'
+#' `letters[grep("[a-m]", letters)]` and `letters[grepl("[a-m]", letters)]`
+#'   both return the same thing, but more circuitously and more verbosely.
+#'
+#' The `stringr` package also provides an even more readable alternative,
+#'   namely `str_subset()`, which should be preferred to versions using
+#'   `str_detect()` and `str_which()`.
+#'
+#' @section Exceptions:
+#'   Note that `x[grep(pattern, x)]` and `grep(pattern, x, value = TRUE)`
+#'   are not _completely_ interchangeable when `x` is not character
+#'   (most commonly, when `x` is a factor), because the output of the
+#'   latter will be a character vector while the former remains a factor.
+#'   It still may be preferable to refactor such code, as it may be faster
+#'   to match the pattern on `levels(x)` and use that to subset instead.
+#'
+#' @evalRd rd_tags("regex_subset_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+regex_subset_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$xml_parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    parent_expr_cond <- xp_and(
+      "OP-LEFT-BRACKET",
+      # parent::expr for LEFT_ASSIGN and RIGHT_ASSIGN, but, strangely,
+      #   parent::equal_assign for EQ_ASSIGN. So just use * as a catchall.
+      "not(parent::*[LEFT_ASSIGN or EQ_ASSIGN or RIGHT_ASSIGN])"
+    )
+    # However the equality operator in XPath works for expr[.] = expr/expr[.],
+    #   i.e., node1 = node2, suits our purpose -- we need that the
+    #   xml2::xml_text() that would come out from the matched nodes is the same,
+    #   i.e. that whatever expression comes in <expr>[grepl(pattern, <expr>)]
+    #   matches exactly, e.g. names(x)[grepl(pattern, names(x))].
+    subset_cond_fmt <- xp_and(
+      "expr[SYMBOL_FUNCTION_CALL[%s]]",
+      "expr[position() = %d] = parent::expr/expr[1]"
```

Comment:
Cool! What is the XPath definition of `=` in this context?
It must ignore attributes I see (col1 e.g. will differ), anything else?

Reply:
I genuinely don't know. would be great to find documentation on this... thus far it's "just worked" wherever needed though (also tested in a few other linters, not sure if any are included yet)

Reply:
Thanks for the spur to check again... found some docs!

https://www.w3.org/TR/1999/REC-xpath-19991116/#booleans

Courtesy this SO Q&A: https://stackoverflow.com/a/4629660/3576984

Reply:
Added that link to the comment